src/garage/torch/algos/maml/maml_base.py

```diff
@@ -0,0 +1,487 @@
+
+    def train_once(self, itr, all_samples, all_params):
+        """Train the algorithm once.
+
+        Args:
+            itr (int): Iteration number.
+            all_samples (list[list[MAMLTrajectoryBatch]]): A two
+                dimensional list of MAMLTrajectoryBatch of size
+                [meta_batch_size * (num_grad_updates + 1)]
+            all_params (list[dict]): A list of named parameter dictionaries.
+                Each dictionary contains key value pair of names (str) and
+                parameters (torch.Tensor).
+
+        Returns:
+            float: Average return.
+
+        """
+        old_theta = dict(self._policy.named_parameters())
+
+        kl_before = self._compute_kl_constraint(itr,
+                                                all_samples,
+                                                all_params,
+                                                set_grad=False)
+
+        meta_objective = self._compute_meta_loss(itr, all_samples, all_params)
+
+        self._meta_optimizer.zero_grad()
+        meta_objective.backward()
+
+        self._meta_optimize(itr, all_samples, all_params)
+
+        # Log
+        loss_after = self._compute_meta_loss(itr,
+                                             all_samples,
+                                             all_params,
+                                             set_grad=False)
+        kl_after = self._compute_kl_constraint(itr,
+                                               all_samples,
+                                               all_params,
+                                               set_grad=False)
+
+        with torch.no_grad():
+            policy_entropy = self._compute_policy_entropy(
+                [task_samples[0] for task_samples in all_samples])
+            average_return = self.evaluate_performance(
+                itr, all_samples, meta_objective.item(), loss_after.item(),
+                kl_before.item(), kl_after.item(),
+                policy_entropy.mean().item())
+
+        update_module_params(self._old_policy, old_theta)
+
+        return average_return
+
+    def _obtain_samples(self, runner):
+        """Obtain samples for each task before and after the fast-adaptation.
+
+        Args:
+            runner (LocalRunner): A local runner instance to obtain samples.
+
+        Returns:
+            tuple: Tuple of (all_samples, all_params).
+                all_samples (list[MAMLTrajectoryBatch]): A list of size
+                    [meta_batch_size * (num_grad_updates + 1)]
+                all_params (list[dict]): A list of named parameter
+                    dictionaries.
+
+        """
+        tasks = self._env.sample_tasks(self._meta_batch_size)
+        all_samples = [[] for _ in range(len(tasks))]
+        all_params = []
+        theta = dict(self._policy.named_parameters())
+
+        for i, task in enumerate(tasks):
+            self._set_task(runner, task)
+            self._set_baseline(i)
```

Comment:
i thought you said that all tasks share a baseline? why do you need multiple baselines?